### PR TITLE
Rebase 1.7.6

### PR DESCRIPTION
**Pull request by soltysh**
@deads2k @liggitt @ironcladlou

### CODE REVIEW

**Comment by soltysh** [comment]
The only problem I have right now is:
```
F0929 16:56:13.334553    6360 openapi.go:38] Failed to register open api spec for root: Duplicate Operation ID createNamespacedDeploymen
tConfigRollback for path /oapi/v1/namespaces/{namespace}/deploymentconfigrollbacks and /oapi/v1/namespaces/{namespace}/deploymentconfigs/{name}/rollback.
```

**Comment by deads2k** [comment]
> The only problem I have right now is:
> 
> F0929 16:56:13.334553    6360 openapi.go:38] Failed to register open api spec for root: Duplicate Operation ID createNamespacedDeploymen
> tConfigRollback for path /oapi/v1/namespaces/{namespace}/deploymentconfigrollbacks and /oapi/v1/namespaces/{namespace}/deploymentconfigs/{name}/rollback.

That's just because your server isn't far enough to fail on something else :)

I think we had a patch somewhere renaming those.

**Comment by soltysh** [comment]
@deads2k found the problem, this was introduced in this fix https://github.com/kubernetes/kubernetes/commit/5b7783f1cbc4df062ca6cb849a5cc883676c9bdf after I removed it it's working. We'll need to figure out if the fix is wrong or we messed something up. I haven't checked how this works in 1.8, though. Let's see how tests results are now.

**Comment by soltysh** [comment]
@liggitt remember when you've mentioned k8s 1.7 bumped azure driver, now I see this is biting us really hard, docker distribution's azure driver relies on some methods not present anymore in the newer version :/ @smarterclayton fyi. What should we do with it?

**Comment by liggitt** [comment]
This is why we *have* to get our dockerregistry built separately, so we can track kube levels of docker, etc. Diamond dependencies are going to become impossible

**Comment by soltysh** [comment]
Yeah, that's the long term goal. But thankfully I realized we have a vendor directory inside our docker/distribution, which just needs the old Azure dependency and it looks like it's working...

**Comment by soltysh** [comment]
This is almost green, except for `TestNodeAuthorizer` and `TestScopedProjectAccess` integration tests.

**Comment by soltysh** [comment]
So apparently rebasing the rebase of the master helped. This is ready for reviews.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/openapi/openapi.go [hunk 1]
```diff
@@ -63,6 +63,10 @@ func GetOperationIDAndTags(r *restful.Route) (string, []string, error) {
 	op := r.Operation
 	path := r.Path
 	var tags []string
+	// TODO: This is hacky, figure out where this name conflict is created and fix it at the root.
+	if strings.HasPrefix(path, "/apis/extensions/v1beta1/namespaces/{namespace}/") && strings.HasSuffix(op, "ScaleScale") {
```
Comment (mfojtik) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/openapi/openapi.go]: create issue pls, this is ultraugly

**Comment by mfojtik** [comment]
Origin changes looks reasonable to me (except https://github.com/openshift/origin/pull/16615#discussion_r142121393 maybe ;-). 

LGTM

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/openapi/openapi.go (hunk 1 above)
Comment (soltysh) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/openapi/openapi.go]: Created https://github.com/openshift/origin/issues/16635.

Comment (deads2k) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/openapi/openapi.go]: mehdy?

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/installer.go [hunk 2]
```diff
@@ -589,15 +589,6 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
 
 		routes := []*restful.RouteBuilder{}
 
-		// If there is a subresource, kind should be the parent's kind.
-		if hasSubresource {
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/installer.go]: I don't think we should do this.  This is messing with the kinds reported via discovery, right?

**Comment by deads2k** [comment]
`UPSTREAM: docker/distribution: <carry>: Azure dependencies` smells like a bad bump.  YOu accidentally wrote a directory should not have perhaps?

**Comment by soltysh** [comment]
> UPSTREAM: docker/distribution: <carry>: Azure dependencies smells like a bad bump. YOu accidentally wrote a directory should not have perhaps?

Nope, @liggitt warned me that 1.7.x bumps Azure deps, which in-turn conflict with docker distrubution's requirements. I had to move the old Azure deps to distribution's vendor dir. It's intentional.

File: Godeps/Godeps.json [hunk 3]
```diff
@@ -1115,317 +1115,322 @@
 		},
 		{
 			"ImportPath": "github.com/docker/distribution",
-			"Comment": "v2.6.2",
-			"Rev": "48294d928ced5dd9b378f7fd7c6f5da3ff3f2c89"
+			"Comment": "v2.6.2-7-g729840c",
+			"Rev": "1e2bbed6e09c6c8047f52af965a6e301f346d04e"
```
Comment (liggitt) [inline Godeps/Godeps.json]: a bump commit consisting only of a sha change doesn't seem correct... there should be file changes associated with this, right?

Comment (soltysh) [reply Godeps/Godeps.json]: These are bumps that match the synced repos in our forks, that David didn't push to origin.

**Comment by soltysh** [comment]
@deads2k whoa, now looking at the openapi diff, it's huge. 👍 for forcing me to fix it properly :)

**Comment by mfojtik** [comment]
test-cmd: etcdserver: request timed out

/retest

**Comment by soltysh** [comment]
[1.7.x rebase](https://trello.com/c/rDJO7CQT/1049-17x-rebase)

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 4]
```diff
@@ -269,7 +269,7 @@ type preparedGenericAPIServer struct {
 // PrepareRun does post API installation setup steps.
 func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
 	if s.swaggerConfig != nil {
-		routes.Swagger{Config: s.swaggerConfig}.Install(s.Handler.GoRestfulContainer)
+		routes.Swagger{Config: s.swaggerConfig}.Install(s.SwaggerAPIContainers(), s.Handler.GoRestfulContainer)
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go]: I assuming something happened in the picklist and you accidentally dropped this

**Comment by deads2k** [comment]
This is dropping fixes like #16392

Not sure how or why. Probably need to resync vendor, re-build the pick list, pick the extras into the branch, and rebump.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go (hunk 4 above)
Comment (soltysh) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go]: Different levels of emicklei/go-restful-swagger12 (iirc) that was failing compilation in our fork. That's why I had to drop it there and re-apply it here.

**Comment by soltysh** [comment]
> This is dropping fixes like #16392

It's there, I'll link to the new commit after I re-push.

> Not sure how or why. Probably need to resync vendor, re-build the pick list, pick the extras into the branch, and rebump.

Some picks were performed after we've had the k8s branch ready and were applied directly in this PR.

**Comment by soltysh** [comment]
@deads2k additionally some commits were dropped since they were aplication+revert pairs. See @liggitt comments on https://github.com/openshift/kubernetes/pull/42/

**Comment by soltysh** [comment]
Additionally, I'm planning to do another round of cherry-picks from 1.7 branch once I'm done with this.

**Comment by deads2k** [comment]
> Additionally, I'm planning to do another round of cherry-picks from 1.7 branch once I'm done with this.

I just resynced the release-1.7.0 branch in openshift/kubernetes

**Comment by soltysh** [comment]
> I just resynced the release-1.7.0 branch in openshift/kubernetes

I'm talking about 1.7 k8s, not our fork.

**Comment by deads2k** [comment]
It failed on `--- FAIL: TestScopedProjectAccess ` twice now

**Comment by soltysh** [comment]
> It failed on --- FAIL: TestScopedProjectAccess twice now

I'll keep my eye on it.

**Comment by soltysh** [comment]
LOL, and I thought that I'm the only one hitting:
```
/usr/lib/golang/pkg/tool/linux_amd64/link: flushing /tmp/go-link-610563765/go.o: write /tmp/go-link-610563765/go.o: cannot allocate memory
```
apparently our CI suffered that one as well.

**Comment by soltysh** [comment]
Failed on:
```
[WARNING] `go test` had the following output to stderr:
# testmain
/usr/lib/golang/pkg/tool/linux_amd64/link: running gcc failed: exit status 1
/usr/bin/ld: final link failed: No space left on device
collect2: error: ld returned 1 exit status

go build testmain: /usr/lib/golang/pkg/tool/linux_amd64/link: signal: killed
# testmain
2017/10/05 07:26:04 reading input: read /go/src/github.com/openshift/origin/_output/local/go/pkg/linux_amd64_race/github.com/openshift/origin/vendor/k8s.io/apimachinery/pkg/util/yaml.a: bad address
# testmain
/usr/lib/golang/pkg/tool/linux_amd64/link: flushing /tmp/go-link-276328245/go.o: write /tmp/go-link-276328245/go.o: cannot allocate memory
# testmain
/usr/lib/golang/pkg/tool/linux_amd64/link: flushing /tmp/go-link-585832900/go.o: write /tmp/go-link-585832900/go.o: cannot allocate memory
# testmain
/usr/lib/golang/pkg/tool/linux_amd64/link: running gcc failed: exit status 1
/usr/bin/ld: final link failed: No space left on device
collect2: error: ld returned 1 exit status
```

**Comment by soltysh** [comment]
`TestScopedProjectAccess` was mentioned by @smarterclayton to be more flaky. I think we'll have to dig into it more.

**Comment by kargakis** [comment]
@deads2k add a retest-not-required label if you want it merged as quick as possible?

**Comment by deads2k** [comment]
> @deads2k add a retest-not-required label if you want it merged as quick as possible?

That carries a risk of breakage, but I can live with that.  Do we have an easily accessible view of the branch jobs to see what happens after this merges?

**Comment by kargakis** [comment]
https://deck-ci.svc.ci.openshift.org/?type=postsubmit

These are jobs that run on every commit in master.

**Comment by deads2k** [comment]
> https://deck-ci.svc.ci.openshift.org/?type=postsubmit
> 
> These are jobs that run on every commit in master.

thanks.

**Comment by kargakis** [comment]
You likely want origin only: https://deck-ci.svc.ci.openshift.org/?repo=openshift%2Forigin&type=postsubmit

Btw, I verified that this PR is not conflicting with the current batch in the queue.

**Comment by deads2k** [comment]
@soltysh failed again

/retest

**Comment by deads2k** [comment]
@soltysh @ironcladlou I need SHAs to start the apimachinery, client-go, apiserver, and metrics release-1.7.6 branches from

**Comment by smarterclayton** [comment]
Godeps references commit "a08f5eeb6246134f4ae5443c0593d72fd057ea7c" which doesn't exist in openshift/kubernetes@a08f5eeb6246134f4ae5443c0593d72fd057ea7c.  Means I can't godep restore - someone planning on fixing this?

**Comment by smarterclayton** [comment]
Another hack/restore-godeps failure:

```
godep: error restoring dep (github.com/docker/distribution/vendor/google.golang.org/grpc/codes): Wanted to restore rev 48294d928ced5dd9b378f7fd7c6f5da3ff3f2c89, already restored rev 1e2bbed6e09c6c8047f52af965a6e301f346d04e for another package in the repo
godep: error restoring dep (github.com/docker/distribution/vendor/google.golang.org/grpc/credentials): Wanted to restore rev 48294d928ced5dd9b378f7fd7c6f5da3ff3f2c89, already restored rev 1e2bbed6e09c6c8047f52af965a6e301f346d04e for another package in the repo
godep: error restoring dep (github.com/docker/distribution/vendor/google.golang.org/grpc/grpclog): Wanted to restore rev 48294d928ced5dd9b378f7fd7c6f5da3ff3f2c89, already restored rev 1e2bbed6e09c6c8047f52af965a6e301f346d04e for another package in the repo
godep: error restoring dep (github.com/docker/distribution/vendor/google.golang.org/grpc/internal): Wanted to restore rev 48294d928ced5dd9b378f7fd7c6f5da3ff3f2c89, already restored rev 1e2bbed6e09c6c8047f52af965a6e301f346d04e for another package in the repo
godep: error restoring dep (github.com/docker/distribution/vendor/google.golang.org/grpc/metadata): Wanted to restore rev 48294d928ced5dd9b378f7fd7c6f5da3ff3f2c89, already restored rev 1e2bbed6e09c6c8047f52af965a6e301f346d04e for another package in the repo
godep: error restoring dep (github.com/docker/distribution/vendor/google.golang.org/grpc/naming): Wanted to restore rev 48294d928ced5dd9b378f7fd7c6f5da3ff3f2c89, already restored rev 1e2bbed6e09c6c8047f52af965a6e301f346d04e for another package in the repo
godep: error restoring dep (github.com/docker/distribution/vendor/google.golang.org/grpc/transport): Wanted to restore rev 48294d928ced5dd9b378f7fd7c6f5da3ff3f2c89, already restored rev 1e2bbed6e09c6c8047f52af965a6e301f346d04e for another package in the repo
godep: error restoring dep (github.com/docker/distribution/vendor/google.golang.org/api/gensupport): Wanted to restore rev 48294d928ced5dd9b378f7fd7c6f5da3ff3f2c89, already restored rev 1e2bbed6e09c6c8047f52af965a6e301f346d04e for another package in the repo
godep: error restoring dep (github.com/docker/distribution/vendor/google.golang.org/appengine/internal/base): Wanted to restore rev 48294d928ced5dd9b378f7fd7c6f5da3ff3f2c89, already restored rev 1e2bbed6e09c6c8047f52af965a6e301f346d04e for another package in the repo
godep: error restoring dep (github.com/docker/distribution/vendor/google.golang.org/appengine/internal/datastore): Wanted to restore rev 48294d928ced5dd9b378f7fd7c6f5da3ff3f2c89, already restored rev 1e2bbed6e09c6c8047f52af965a6e301f346d04e for another package in the repo
```

Godeps/Godeps.json has docker/distribution in there twice at two different Revs.

**Comment by ironcladlou** [comment]
So this is weird....

The referenced commit merged to release-1.7.6 from my original rebase PR in https://github.com/openshift/kubernetes/pull/42/commits/a08f5eeb6246134f4ae5443c0593d72fd057ea7c: "soltysh  merged 166 commits into openshift:release-1.7.6 from ironcladlou:rebase-1.7.6"

Notice that https://github.com/openshift/kubernetes/commit/a08f5eeb6246134f4ae5443c0593d72fd057ea7c isn't in the `release-1.7.6` branch now.

BUT, here's a duplicate of the commit with a different sha in the `release-1.7.6` branch https://github.com/openshift/kubernetes/commit/ad0dbed99c9ad048da3d493f90e3e1847f35c92c.

👀

**Comment by soltysh** [comment]
@deads2k and I were syncing our forks, either of us must have overwritten the commits. Looking at the commits it appears it was me, not sure why though, since I was working on top of freshly pulled branch...

**Pull request merged**